### PR TITLE
feat: Added persistent segment tree with lazy propagation

### PR DESCRIPTION
**Pull request by MSedra**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1445"></a>

### CODE REVIEW

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Code is not up to the [repository standards](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md).
Please read them **carefully** and **follow them**.

**Comment by MSedra** [comment]
> Code is not up to the [repository standards](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md).
> Please read them **carefully** and **follow them**.

I have read the standards and tried to follow them as much as i could. Can you please elaborate which standard/s am i not following? Is using normal pointers not allowed?

**Review by Panquesito7 — COMMENTED** [review]
Let's start: first, we need to fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1445/checks?check_run_id=1778864629) warnings. 🙂

**Comment by MSedra** [comment]
> Let's start: first, we need to fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1445/checks?check_run_id=1778864629) warnings. 🙂

Why do i get warnings when using normal pointers?

**Comment by MSedra** [comment]
> Let's start: first, we need to fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1445/checks?check_run_id=1778864629) warnings. 🙂

Done. What's next?

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/MSedra/C-Plus-Plus/actions

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
+/**
+ * @file
+ * @brief [Persistent segment tree with range updates (lazy propagation)](https://en.wikipedia.org/wiki/Persistent_data_structure - https://www.geeksforgeeks.org/lazy-propagation-in-segment-tree/)
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:3]: This is Markdown, and it won't work properly if you add two links like you did. Please choose one.

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+ * It saves all previous versions of itself (before every update) to facilitate doing range queries in any version.
+ * More memory is used ,but the logarithmic time is preserved because the new version points to the same nodes, that the previous version points to, that are not affected by the update.
+ * That is, only the nodes that are affected by the update and their ancestors are copied. The rest is copied using lazy propagation in the next queries.
+ * Thus preserving the logarithmic time because the number of nodes copied after any update is logarithmic.
+ *
+ * @author [Magdy Sedra](https://github.com/MSedra)
+ */
+#include<iostream>
+#include<vector>
+#include <memory> /// to manage dynamic memory
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:20]:
```suggestion
#include <iostream>  /// for IO operations
#include <vector>    /// for std::vector
#include <memory>    /// to manage dynamic memory
```

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    }
+
+    /**
+     * @brief If there is some value to be propagated to the passed node, value is added to the node and the children of the node, if exist, are copied and the propagated value is also added to them
+     * @param i the left index of the range that the passed node holds its sum
+     * @param j the right index of the range that the passed node holds its sum
+     * @param curr pointer to the node to be propagated
+     * @returns void
+     */
+    void lazy(int i, int j, std::shared_ptr<Node> const &curr) {
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:62]: As these two variables are not being modified, consider making them a `const` and pass them by reference.
```suggestion
    void lazy(const int& i, const int& j, std::shared_ptr<Node> const &curr) {
```

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        curr->prop = 0;
+    }
+
+    /**
+     * @brief Constructing the segment tree with the early passed vector. Every call creates a node to hold the sum of the given range, set its pointers to the children, and set its value to the sum of the children's values
+     * @param i the left index of the range that the created node holds its sum
+     * @param j the right index of the range that the created node holds its sum
+     * @returns pointer to the newly created node
+     */
+    std::shared_ptr<Node> construct(int i, int j) {
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:82]:
```suggestion
    std::shared_ptr<Node> construct(const int& i, const int& j) {
```

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+     * @brief Doing range update, checking at every node if it has some value to be propagated. All nodes affected by the update are copied and propagation value is added to the leaf of them
+     * @param i the left index of the range that the passed node holds its sum
+     * @param j the right index of the range that the passed node holds its sum
+     * @param l the left index of the range to be updated
+     * @param r the right index of the range to be updated
+     * @param value the value to be added to every element whose index x satisfies l<=x<=r
+     * @param curr pointer to the current node, which has value = the sum of elements whose index x satisfies i<=x<=j
+     * @return
+     */
+    std::shared_ptr<Node> update(int i, int j, int l, int r, int value, std::shared_ptr<Node> const &curr) {
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:107]:
```suggestion
    std::shared_ptr<Node> update(const int& i, const int& j, const int& l, const int& r, const int& value, std::shared_ptr<Node> const &curr) {
```

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    /**
+     * @brief Querying the range from index l to index r, checking at every node if it has some value to be propagated. Current node's value is returned if its range is completely inside the wanted range, else 0 is returned
+     * @param i the left index of the range that the passed node holds its sum
+     * @param j the right index of the range that the passed node holds its sum
+     * @param l the left index of the range whose sum should be returned as a result
+     * @param r the right index of the range whose sum should be returned as a result
+     * @param curr pointer to the current node, which has value = the sum of elements whose index x satisfies i<=x<=j
+     * @return sum of elements whose index x satisfies l<=x<=r
+     */
+    int64_t query(int i, int j, int l, int r, std::shared_ptr<Node> const &curr) {
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:135]:
```suggestion
    int64_t query(const int& i, const int& j, const int& l, const int& r, std::shared_ptr<Node> const &curr) {
```

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+     * @brief Getting the number of versions after updates so far which is equal to the size of the pointers vector
+     * @return the number of versions
+     */
+    int size() // returns the number of segment trees (versions) , the number of updates done so far = returned value - 1 ,because one of the trees is the original segment tree
+    {
+        return ptrs.size();
+    }
+};
+
+void test() {
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:202]:
```suggestion
/**
 * @brief Test implementations
 * @returns void
 */
static void test() {
```

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+            std::cout << ",";
+        }
+    }
+    std::cout << "}\n";
+    std::cout << "Number of segment trees (versions) now = " << tree.size() << '\n';
+    std::cout << "Querying range sum on version 0 from index 3 to 5 = 11-2+7 = " << tree.query(3, 5, 0) << '\n';
+    std::cout << "Querying range sum on version 1 from index 3 to 5 = 4-9+0 = " << tree.query(3, 5, 1) << '\n';
+}
+
+int main() {
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:240]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
```

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+    test();
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:241]:
```suggestion
    test();  // run self-test implementations
```

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+    return 0;
+}
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:243]:
```suggestion
}

```

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+ * That is, only the nodes that are affected by the update and their ancestors are copied. The rest is copied using lazy propagation in the next queries.
+ * Thus preserving the logarithmic time because the number of nodes copied after any update is logarithmic.
+ *
+ * @author [Magdy Sedra](https://github.com/MSedra)
+ */
+#include<iostream>
+#include<vector>
+#include <memory> /// to manage dynamic memory
+
+/// Range query here is range sum ,but the code can be modified to make different queries like range max or min
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:22]:
```suggestion
/**
 * @namespace range_queries
 * @brief Range queries algorithms
 */
namespace range_queries {
/**
 * @brief Range query here is range sum, but the code can be
 modified to make different queries like range max or min.
 */
```

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+
+    /**
+     * @brief Getting the number of versions after updates so far which is equal to the size of the pointers vector
+     * @return the number of versions
+     */
+    int size() // returns the number of segment trees (versions) , the number of updates done so far = returned value - 1 ,because one of the trees is the original segment tree
+    {
+        return ptrs.size();
+    }
+};
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:200]:
```suggestion
};
}  // namespace range_queries
```

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+
+void test() {
+    std::vector<int> arr = {-5, 2, 3, 11, -2, 7, 0, 1};
+    perSegTree tree;
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:204]:
```suggestion
    range_queries::perSegTree tree;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 14]
```diff
@@ -0,0 +1,323 @@
+/**
+ * @file
+ * @brief [Persistent segment tree with range updates (lazy
+ * propagation)](https://en.wikipedia.org/wiki/Persistent_data_structure)
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:4]:
```suggestion
 * propagation)](https://en.wikipedia.org/wiki/Persistent_data_structure)

```

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+                            /// prop is the value to be propagated/added to all
+                            /// the leaf nodes children of that node
+
+        Node() {
+            val = prop = 0;
+            left = right = nullptr;
+        }
+    };
+
+    int n;  /// number of elements/leaf nodes in the segment tree
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:55]: We **always** got to initialize all variables, at least with `0`.
```suggestion
    int n = 0;  /// number of elements/leaf nodes in the segment tree
```

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+    std::vector<std::shared_ptr<Node>>
+        ptrs;  /// ptrs[i] holds a root pointer to the segment tree after the
+               /// ith update. ptrs[0] holds a root pointer to the segment tree
+               /// before any updates
+    std::vector<int> vec;  /// values of the leaf nodes that the segment tree
```
Comment (Panquesito7) [inline range_queries/persistent_segment_tree_with_lazy_propagation.cpp:60]:
```suggestion
    std::vector<int> vec{};  /// values of the leaf nodes that the segment tree
```

**Review by Panquesito7 — COMMENTED** [review]
👍 Nice work! Well done.
Would you please consider editing the filename to make it shorter?
Also consider using `uint64_t` (for non-negative values), `int64_t` (for negative values) or their appropriate size in all the code.

Thank you. 🙂

File: range_queries/persistent_segment_tree_with_lazy_propagation.cpp (hunk 14 above)
Comment (MSedra) [reply range_queries/persistent_segment_tree_with_lazy_propagation.cpp]: Done. 

Thank you so much.

**Comment by MSedra** [comment]
@Panquesito7 Can you please rereview?

**Comment by MSedra** [comment]
@ayaankhan98 can you please review the last changes?

**Review by ayaankhan98 — APPROVED** [review]

**Review by Panquesito7 — APPROVED** [review]
Great work, looks amazing! LGTM; thanks for your contribution! 👍 🎉

**Comment by MSedra** [comment]
> Great work, looks amazing! LGTM; thanks for your contribution! 👍 🎉

Thank you so much for your help

**Pull request merged**